#######################################################
#
#	selectAssociatedGenes.R
#	SELECTS GENES NEAR TO X MB OF A SNP ASSOCIATED
#	WITH A DETERMINED PHENOTYPE.
#
#	AUTHOR. JAVIER MONTALVO-ARREDONDO.
#	VERSION. 0
#	CONTACT. buitrejma[at]gmail[dot]com
#	CALZADA ANTONIO NARRO, 1923,
#	SAN BUENAVISTA SALTILLO COAHUILA, C.P. 25315
#
#	TAKES THE GWAS ANALYSIS RESULTS TABLE, A TRANSCRIPTOME ANNOTATION
#	FILE IN GFF FORMAT AND A HEADER HASH TABLE RELATING CHROMOSOME
#	NAME AND NUMBER OF CHROMOSOME. THIS CODE WILL SELECT GENES NEAR 
#	TO A SIGNIFICATIVE SNP ASSOCIATED WITH PHENOTYPE.
#
#
#	PSEUDO-CODE.
#	
#	10	DECLARE XMB AS INT: NUMBER IN MEGABASES TO SEARCH GENES NEAR TO THE SIGNIFICATIVE SNP
#	15	DECLARE PROD_CONTAINER AS A MATRIX WITH DIMS OF 1 X 4 
#	20	WITH INPUT FILES CREATE DATA OBJECTS (OBJECT-GWAS-RESULTS, OBJECT-HEADER-HASH, OBJECT-GFF).
#	30	ITERATE AMONG OBJECT-GWAS-RESULTS ROWS.
#	3010		LET NAME  TAKES OBJECT-GWAS-RESULTS$SNP_ID.
#	3020		LET CHRN  TAKES CHROMOSOME_NAME FROM OBJECT-HEADER-HASH USING OBJECT-GWAS-RESULTS$CHR_NUMBER AS A KEY.
#	3030		LET POS   TAKES OBJECT-GWAS-RESULTS$SNP_POSITION.
#	3035		LET PVAL  TAKES OBJECT-GWAS-RESULTS$SNP_PVLAUE.
#	3040		LET START TAKES CALCULATION OF POS - XMB.
#	3050		LET END   TAKES CALCULATION OF POS + XMB.
#	3060		LET TMP-OBJECT-GFF TAKES SAMPLED OBJECT-GFF BY CHRN
#	3070		LET PROD  TAKES TMP-OBJECT-GFF$PRODUCT IF [TMP-OBJECT-GFF$START >= START & TMP-OBJECT-GFF$END <= END] OR DO CONTINUE
#	3075		LET NAME, CHRN, POS BE THE SAME LENGTH OF A PROD COLUMN.
#	3080		LET TMP_PROD_CONTAINER BE A DATAFRAME
#	3090		LET TMP_PROD_CONTAINER CONTAINS NAME, CHRN, POS, PROD.
#	40	RETURN PROD_CONTAINER
############################################################


selectAssociatedGenes <- function(xmb, input_files) {
	#	LOADING LIBRARY
	stopifnot("rtracklayer" %in% rownames(installed.packages()))
	library(rtracklayer)
	#	CREATING OBJECTS FROM DATA
	GWAS <- read.delim(input_files[1], header=TRUE, sep=",", as.is=TRUE)
	GFF <- import.gff3(input_files[2])
	HEADER <- read.delim(input_files[3], header=FALSE, sep="|", as.is=TRUE)

	num_of_gwas_rows <- length(GWAS[,1])

	for (i in 1:num_of_gwas_rows) {
		name <- GWAS[i,1]
		chrn <- HEADER[as.numeric(GWAS[i,2]) == chrn,1]
		pos <- as.numeric(GWAS[i,3])
		pval <- as.numeric(GWAS[i,4]) # I do not remeber what the column number is. Watch Out!
		start <- pos - xmb
		end <- pos + xmb
		
	}
}
